CA2021 false positive on Enumerable.OfType<T>extension

### Analyzer
**Diagnostic ID**: [CA2021](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2021): `CA2021: Don't call Enumerable.Cast<T> or Enumerable.OfType<T> with incompatible types`

### Analyzer source
**SDK**: [Built-in CA analyzers in .NET 8 SDK or later](https://learn.microsoft.com/dotnet/fundamentals/productivity/code-analysis)
**Version**: [SDK 8.0.101](https://dotnet.microsoft.com/download/dotnet/8.0)

### Describe the bug
Warning CA2021 is found on Enumerable.OfType<T> call when T is compatible with underlining type in the enumerable collection.

### Steps To Reproduce
```
	public static void Main()
	{
		var nodeChanges = new List<INodeUpdate<GraphNode>> { new DataNodeUpdate(new DataNode(0, 0)) };

		//warning CA2021: This call will always result in an empty sequence because type 'INodeUpdate<GraphNode>' is incompatible with type 'DataNodeUpdate'
		var nodeChangesFiltered = nodeChanges.OfType<DataNodeUpdate>();

		Debug.Assert(nodeChangesFiltered.Count() == 1);
	}

	public abstract record class GraphNode(long Id);
	public sealed record class DataNode(long Id, long Value) : GraphNode(Id);

	public interface INodeUpdate<out T>
	{
		T Updated { get; }
	}

	public abstract record class NodeUpdate<T>(T Updated) : INodeUpdate<T> where T : GraphNode;
	public sealed record class DataNodeUpdate(DataNode Updated) : NodeUpdate<DataNode>(Updated);
```

### Expected behavior
No CA2021 warning.

### Actual behavior
CA2021 warning is found.

--- a/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastOrOfTypeWithIncompatibleTypesAnalyzer.cs
+++ b/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastOrOfTypeWithIncompatibleTypesAnalyzer.cs
@@ -196,6 +196,9 @@ namespace Microsoft.NetCore.Analyzers.Runtime
                     return false;
                 }
 
+                castFrom = castFrom.OriginalDefinition;
+                castTo = castTo.OriginalDefinition;
+
                 if (castFrom.SpecialType == SpecialType.System_Object
                    || castTo.SpecialType == SpecialType.System_Object)
                 {
@@ -220,14 +223,14 @@ namespace Microsoft.NetCore.Analyzers.Runtime
                 //   UnmanagedTypeConstraint
                 //   Nullability annotations
 
-                switch (castFrom.OriginalDefinition.TypeKind, castTo.OriginalDefinition.TypeKind)
+                switch (castFrom.TypeKind, castTo.TypeKind)
                 {
                     case (TypeKind.Dynamic, _):
                     case (_, TypeKind.Dynamic):
                         return false;
 
                     case (TypeKind.TypeParameter, _):
-                        var castFromTypeParam = (ITypeParameterSymbol)castFrom.OriginalDefinition;
+                        var castFromTypeParam = (ITypeParameterSymbol)castFrom;
                         if (IsUnconstrainedTypeParameter(castFromTypeParam))
                         {
                             return false;
@@ -246,7 +249,7 @@ namespace Microsoft.NetCore.Analyzers.Runtime
 
                         return false;
                     case (_, TypeKind.TypeParameter):
-                        var castToTypeParam = (ITypeParameterSymbol)castTo.OriginalDefinition;
+                        var castToTypeParam = (ITypeParameterSymbol)castTo;
                         if (IsUnconstrainedTypeParameter(castToTypeParam))
                         {
                             return false;
@@ -270,32 +273,32 @@ namespace Microsoft.NetCore.Analyzers.Runtime
                             && !castTo.DerivesFrom(castFrom);
 
                     case (TypeKind.Interface, TypeKind.Class):
-                        return castTo.IsSealed && !castTo.AllInterfaces.Contains(castFrom);
+                        return castTo.IsSealed && !castTo.DerivesFrom(castFrom);
                     case (TypeKind.Class, TypeKind.Interface):
-                        return castFrom.IsSealed && !castFrom.AllInterfaces.Contains(castTo);
+                        return castFrom.IsSealed && !castFrom.DerivesFrom(castTo);
 
                     case (TypeKind.Interface, TypeKind.Struct):
-                        return !castTo.AllInterfaces.Contains(castFrom);
+                        return !castTo.DerivesFrom(castFrom);
                     case (TypeKind.Struct, TypeKind.Interface):
-                        return !castFrom.AllInterfaces.Contains(castTo);
+                        return !castFrom.DerivesFrom(castTo);
 
                     case (TypeKind.Class, TypeKind.Enum):
-                        return castFrom.OriginalDefinition.SpecialType is not SpecialType.System_Enum
-                                                                      and not SpecialType.System_ValueType;
+                        return castFrom.SpecialType is not SpecialType.System_Enum
+                                                   and not SpecialType.System_ValueType;
                     case (TypeKind.Enum, TypeKind.Class):
-                        return castTo.OriginalDefinition.SpecialType is not SpecialType.System_Enum
-                                                                    and not SpecialType.System_ValueType;
+                        return castTo.SpecialType is not SpecialType.System_Enum
+                                                 and not SpecialType.System_ValueType;
 
                     case (TypeKind.Struct, TypeKind.Enum)
-                    when castTo.OriginalDefinition is INamedTypeSymbol toEnum:
+                    when castTo is INamedTypeSymbol toEnum:
                         return !castFrom.Equals(toEnum.EnumUnderlyingType);
                     case (TypeKind.Enum, TypeKind.Struct)
-                    when castFrom.OriginalDefinition is INamedTypeSymbol fromEnum:
+                    when castFrom is INamedTypeSymbol fromEnum:
                         return !fromEnum.EnumUnderlyingType!.Equals(castTo);
 
                     case (TypeKind.Enum, TypeKind.Enum)
-                    when castFrom.OriginalDefinition is INamedTypeSymbol fromEnum
-                      && castTo.OriginalDefinition is INamedTypeSymbol toEnum:
+                    when castFrom is INamedTypeSymbol fromEnum
+                      && castTo is INamedTypeSymbol toEnum:
                         return !fromEnum.EnumUnderlyingType!.Equals(toEnum.EnumUnderlyingType);
 
                     // this is too conservative
@@ -310,14 +313,14 @@ namespace Microsoft.NetCore.Analyzers.Runtime
                             || CastWillAlwaysFail(fromArray.ElementType, toArray.ElementType);
 
                     case (TypeKind.Array, TypeKind.Class):
-                        return castTo.OriginalDefinition.SpecialType != SpecialType.System_Array;
+                        return castTo.SpecialType != SpecialType.System_Array;
                     case (TypeKind.Class, TypeKind.Array):
-                        return castFrom.OriginalDefinition.SpecialType != SpecialType.System_Array;
+                        return castFrom.SpecialType != SpecialType.System_Array;
 
                     case (TypeKind.Class, TypeKind.Struct):
-                        return castFrom.OriginalDefinition.SpecialType != SpecialType.System_ValueType;
+                        return castFrom.SpecialType != SpecialType.System_ValueType;
                     case (TypeKind.Struct, TypeKind.Class):
-                        return castTo.OriginalDefinition.SpecialType != SpecialType.System_ValueType;
+                        return castTo.SpecialType != SpecialType.System_ValueType;
 
                     case (_, TypeKind.Enum):
                     case (TypeKind.Enum, _):

--- a/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastOrOfTypeWithIncompatibleTypesAnalyzerTests.cs
+++ b/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastOrOfTypeWithIncompatibleTypesAnalyzerTests.cs
@@ -8,6 +8,7 @@ using System.Threading.Tasks;
 using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.CSharp;
 using Microsoft.CodeAnalysis.Testing;
+using Test.Utilities;
 using Xunit;
 using VerifyCS = Test.Utilities.CSharpCodeFixVerifier<
     Microsoft.NetCore.Analyzers.Runtime.DoNotCallEnumerableCastOrOfTypeWithIncompatibleTypesAnalyzer,

--- a/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastOrOfTypeWithIncompatibleTypesAnalyzerTests.cs
+++ b/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastOrOfTypeWithIncompatibleTypesAnalyzerTests.cs
@@ -1,4 +1,4 @@
-// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.
 
 using System;
 using System.Collections.Generic;
@@ -7,6 +7,7 @@ using System.Text.RegularExpressions;
 using System.Threading.Tasks;
 using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Testing;
 using Xunit;
 using VerifyCS = Test.Utilities.CSharpCodeFixVerifier<
     Microsoft.NetCore.Analyzers.Runtime.DoNotCallEnumerableCastOrOfTypeWithIncompatibleTypesAnalyzer,
@@ -823,6 +824,45 @@ class C : IInterface
     VerifyCS.Diagnostic(castRule).WithLocation(56).WithArguments("string", "TStruct?"),
                 }
             }.RunAsync();
+        }
+
+        [Fact]
+        public async Task GenericRecordsAndInterfaces()
+        {
+            var test = new VerifyCS.Test
+            {
+                ReferenceAssemblies = ReferenceAssemblies.Net.Net70,
+                LanguageVersion = LanguageVersion.CSharp10,
+                TestCode = @"
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+
+public static class Program
+{
+    public static void Main()
+    {
+        var nodeChanges = new List<INodeUpdate<GraphNode>> { new DataNodeUpdate(new DataNode(0, 0)) };
+
+        //warning CA2021: This call will always result in an empty sequence because type 'INodeUpdate<GraphNode>' is incompatible with type 'DataNodeUpdate'
+        var nodeChangesFiltered = nodeChanges.OfType<DataNodeUpdate>();
+
+        Debug.Assert(nodeChangesFiltered.Count() == 1);
+    }
+}
+
+public abstract record class GraphNode(long Id);
+public sealed record class DataNode(long Id, long Value) : GraphNode(Id);
+
+public interface INodeUpdate<out T>
+{
+    T Updated { get; }
+}
+
+public abstract record class NodeUpdate<T>(T Updated) : INodeUpdate<T> where T : GraphNode;
+public sealed record class DataNodeUpdate(DataNode Updated) : NodeUpdate<DataNode>(Updated);"
+            };
+            await test.RunAsync();
         }
 
         [Fact]

--- a/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastOrOfTypeWithIncompatibleTypesAnalyzerTests.cs
+++ b/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastOrOfTypeWithIncompatibleTypesAnalyzerTests.cs
@@ -826,7 +826,7 @@ class C : IInterface
             }.RunAsync();
         }
 
-        [Fact]
+        [Fact, WorkItem(7153, "https://github.com/dotnet/roslyn-analyzers/issues/7153")]
         public async Task GenericRecordsAndInterfaces()
         {
             var test = new VerifyCS.Test

